cmd/soroban-cli/src/contract/install.rs

```diff
@@ -96,42 +51,28 @@ impl Cmd {
     }
 
     pub async fn run_and_get_hash(&self) -> Result<String, Error> {
-        let contract = fs::read(&self.wasm).map_err(|e| Error::CannotReadContractFile {
-            filepath: self.wasm.clone(),
-            error: e,
-        })?;
-
-        if self.rpc_url.is_some() {
-            self.run_against_rpc_server(contract).await
-        } else {
+        let contract = self.wasm.read()?;
+        if self.config.is_no_network() {
             self.run_in_sandbox(contract)
+        } else {
+            self.run_against_rpc_server(contract).await
         }
     }
 
     fn run_in_sandbox(&self, contract: Vec<u8>) -> Result<String, Error> {
-        let mut state = utils::ledger_snapshot_read_or_default(&self.ledger_file).map_err(|e| {
-            Error::CannotReadLedgerFile {
-                filepath: self.ledger_file.clone(),
-                error: e,
-            }
-        })?;
+        let mut state = self.config.get_state()?;
         let wasm_hash =
             utils::add_contract_code_to_ledger_entries(&mut state.ledger_entries, contract)?;
 
-        state
-            .write_file(&self.ledger_file)
-            .map_err(|e| Error::CannotCommitLedgerFile {
-                filepath: self.ledger_file.clone(),
-                error: e,
-            })?;
+        self.config.set_state(&mut state)?;
 
         Ok(hex::encode(wasm_hash))
     }
 
     async fn run_against_rpc_server(&self, contract: Vec<u8>) -> Result<String, Error> {
-        let client = Client::new(self.rpc_url.as_ref().unwrap());
-        let key = utils::parse_secret_key(self.secret_key.as_ref().unwrap())
-            .map_err(|_| Error::CannotParseSecretKey)?;
+        let network = self.config.get_network()?;
+        let client = Client::new(&network.rpc_url);
+        let key = self.config.key_pair()?;
```

Comment:
May not have a keypair if we're using, say... a hw signer like a ledger. We will only have the public key. So maybe the methods would be `public_key()`, and `sign(msg)`?

Reply:
But that's a way bigger refactor, so let's leave it for now.

Reply:
Yep.  The same will be true for mac keychain.  Will address this in a future PR, perhaps should be a separate crate.